flash/data/data_module.py

```diff
@@ -0,0 +1,315 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import platform
+from typing import Any, Callable, Optional, Union
+
+import pytorch_lightning as pl
+import torch
+from numpy import isin
+from pytorch_lightning.trainer.states import RunningStage
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from torch.utils.data import DataLoader, Dataset
+from torch.utils.data.dataset import Subset
+
+from flash.data.auto_dataset import AutoDataset
+from flash.data.data_pipeline import DataPipeline, Postprocess, Preprocess
+
+
+class MockLightningModule(pl.LightningModule):
+
+    pass
+
+
+class TaskDataPipeline(DataPipeline):
+
+    def per_batch_transform(self, batch: Any) -> Any:
+        return (batch["x"], batch.get('target', batch.get('y'))) if isinstance(batch, dict) else batch
+
+
+class DataModule(pl.LightningDataModule):
+    """Basic DataModule class for all Flash tasks
+
+    Args:
+        train_ds: Dataset for training. Defaults to None.
+        valid_ds: Dataset for VALIDATING model performance during training. Defaults to None.
+        test_ds: Dataset to test model performance. Defaults to None.
+        batch_size: the batch size to be used by the DataLoader. Defaults to 1.
+        num_workers: The number of workers to use for parallelized loading.
+            Defaults to None which equals the number of available CPU threads,
+            or 0 for Darwin platform.
+    """
+
+    preprocess_cls = Preprocess
+    postprocess_cls = Postprocess
+
+    def __init__(
+        self,
+        train_ds: Optional[AutoDataset] = None,
+        valid_ds: Optional[AutoDataset] = None,
+        test_ds: Optional[AutoDataset] = None,
+        predict_ds: Optional[AutoDataset] = None,
+        batch_size: int = 1,
+        num_workers: Optional[int] = None,
+    ):
+        super().__init__()
+        self._train_ds = train_ds
+        self._valid_ds = valid_ds
+        self._test_ds = test_ds
+        self._predict_ds = predict_ds
+
```

Comment:
shouldn't we raise an error if all passed ds are None ?

Reply:
No, there is also an option of injecting them later on. And sometimes we also need to create an instance here without any data, just to use some of the other properties. This would block us then.

Reply:
ok, then I see here that there's need of two different classes. As a user it doesn't make sense that I can create a DataModule with empty (or maybe wrong path) without noticing that.

Reply:
I disagree here again.

If I want to add my data to the module later on, I should have the option to do so. DataModules should basically just be containers.

Reply:
if it's not here, there's a leak somewhere at least from user experience.

Reply:
I don't think so. I think it's definitely valid to have something like this:

```python

dm = MyDataModule()
dataset = ... # some logic that relies on dm internals
dm.trainset = dataset
```

What we are doing here is just a default, but we don't want to force the user to use it exactly like this.